libraries/AP_BattMonitor/AP_BattMonitor_TIBQ76952.cpp

```diff
@@ -0,0 +1,937 @@
+    _state.voltage = accumulate.voltage / accumulate.count;
+    _state.state_of_health_pct = accumulate.health_pct / accumulate.count;
+    _state.has_state_of_health_pct = true;
+    _state.current_amps = fabsf(accumulate.current / accumulate.count);
+    _state.temperature = accumulate.temp / accumulate.count;
+    for (uint8_t i = 0; i < MIN(HAL_BATTMON_CELL_COUNT, AP_BATT_MONITOR_CELLS_MAX); i++) {
+        _state.cell_voltages.cells[i] = accumulate.cell_voltages_mv[i] / accumulate.count;
+    }
+
+    // clear accumulate structure
+    accumulate = {};
+}
+
+// set desired powered state (enabled/disabled) by enabling/disabling discharge FET
+void AP_BattMonitor_TIBQ76952::mppt_set_powered_state(bool power_on)
+{
+    if (!configured) {
+        return;
+    }
+    sub_command(power_on ? TIBQ769x2_ALL_FETS_ON : TIBQ769x2_ALL_FETS_OFF);
+}
+
+// periodic timer callback
+void AP_BattMonitor_TIBQ76952::timer(void)
+{
+    // configure device if required
+    if (!configure()) {
+        return;
+    }
+
+    // read data from device
+    read_voltage_current_temperature();
+    read_power_state();
+    read_safety_state();
+}
+
+// configure device
+bool AP_BattMonitor_TIBQ76952::configure()
+{
+    // exit immediately if already configured
+    if (configured) {
+        return true;
+    }
+
+    // check device id, exit on failure
+    const uint32_t device_number = sub_command_read_4bytes(TIBQ769x2_DEVICE_NUMBER);
+    if (device_number != DEVICE_ID_TIBQ7695) {
+        Debug("BQ76952: Unknown device detected - ID: 0x%08lX", (unsigned long)device_number);
+        return false;
+    }
+
+    // check device's firmware and hardware versions
+#if DEBUG_PRINT
+    const uint32_t fw_version = sub_command_read_4bytes(TIBQ769x2_FW_VERSION);
+    const uint32_t hw_version = sub_command_read_4bytes(TIBQ769x2_HW_VERSION);
+    Debug("BQ76952 detected, fw: 0x%08lX, hw: 0x%08lX", (unsigned long)fw_version, (unsigned long)hw_version);
+#endif
+
+    // enable REG1 3.3v to provide power to the MCU
+    set_register(TIBQ769x2_REG12Config, 0x01, 1); // 0x01 for 3.3V output
+
+    // wake up device
+    sub_command(TIBQ769x2_EXIT_DEEPSLEEP);
+    hal.scheduler->delay(10);
+    sub_command(TIBQ769x2_SLEEP_DISABLE);
+    hal.scheduler->delay(10);
+
+    // Clear any remaining permanent failure alerts
+    direct_command(TIBQ769x2_PFAlertA, 0xFF, CommandType::WRITE);
+    direct_command(TIBQ769x2_PFAlertB, 0xFF, CommandType::WRITE);
+    direct_command(TIBQ769x2_PFAlertC, 0xFF, CommandType::WRITE);
+    direct_command(TIBQ769x2_PFAlertD, 0xFF, CommandType::WRITE);
+
+    // Enter CONFIGUPDATE mode (Subcommand 0x0090) - Required to program device RAM settings
+    sub_command(TIBQ769x2_SET_CFGUPDATE);
+
+    // 'Power Config' - 0x9234 = 0x2D80
+    // Setting the DSLP_LDO bit allows the LDOs to remain active when the device goes into Deep Sleep mode
+    // Set wake speed bits to 00 for best performance
+    set_register(TIBQ769x2_PowerConfig, 0x2D80, 2);
```

Comment:
These calls should instead be an interation through a `static const` array.

Reply:
ok, added to a big static const array called config_settings